tensorflow_probability/python/experimental/mcmc/gibbs_kernel.py

```diff
@@ -0,0 +1,411 @@
+                                 'y': y})
+            return tf.reduce_sum(lp)
+
+
+        # random walk Markov chain function
+        def kernel_make_fn(target_log_prob_fn, state):
+            return tfp.mcmc.RandomWalkMetropolis(target_log_prob_fn=target_log_prob_fn)
+
+
+        # posterior distribution MCMC chain
+        @tf.function
+        def posterior(iterations, burnin, thinning, initial_state):
+            kernel_list = [(0, kernel_make_fn), # conditional probability for zeroth parmeter alpha
+                           (1, kernel_make_fn), # conditional probability for first parameter beta
+                           (2, kernel_make_fn)] # conditional probability for second parameter sigma
+            kernel = GibbsKernel(
+                target_log_prob_fn=log_prob,
+                kernel_list=kernel_list
+            )
+            return tfp.mcmc.sample_chain(
+                num_results=iterations,
+                current_state=initial_state,
+                kernel=kernel,
+                num_burnin_steps=burnin,
+                num_steps_between_results=thinning,
+                parallel_iterations=1,
+                trace_fn=None)
+
+
+        # initialize model
+        model = Model(x)
+        initial_state = [dtype(0.1), dtype(0.1), dtype(0.1)]  # start chain at alpha=0.1, beta=0.1, sigma=0.1
+
+        # estimate posterior distribution
+        samples = posterior(
+            iterations=10000,
+            burnin=1000,
+            thinning=0,
+            initial_state=initial_state)
+
+        tf.print('alpha samples:', samples[0])
+        tf.print('beta  samples:', samples[1])
+        tf.print('sigma samples:', samples[2])
+        tf.print('sample means: [alpha, beta, sigma] =', tf.math.reduce_mean(samples, axis=1))
+
+    ```
+
+
+
+    """
+
+    def __init__(self, target_log_prob_fn, kernel_list, name=None):
+        """Build a Gibbs sampling scheme from component kernels.
+
+        :param target_log_prob_fn: a function that takes `state` arguments
+                                   and returns the target log probability
+                                   density.
+        :param kernel_list: a list of tuples `(state_part_idx, kernel_make_fn)`.
+                            `state_part_idx` denotes the index (relative to
+                            positional args in `target_log_prob_fn`) of the
+                            state the kernel updates.  `kernel_make_fn` takes
+                            arguments `target_log_prob_fn` and `state`, returning
+                            a `tfp.mcmc.TransitionKernel`.
+        :returns: an instance of `GibbsKernel`
+        """
+        # Require to check if all kernel.is_calibrated is True
+        self._parameters = dict(
+            target_log_prob_fn=target_log_prob_fn,
+            kernel_list=kernel_list,
+            name=name,
+        )
+
+    @property
+    def is_calibrated(self):
+        return True
```

Comment:
presumably only calibrated as long as each make_kernel_fn is also calibrated?

Reply:
@brianwa84 Just a few preliminary comments on the above:
1. That's a really good point, we'd thought about that too.  I'd vote for the `SamplingKernel` approach, as it simplifies `GibbsKernel` and explicitly marks the step out as a full conditional.  How about `FullConditionalKernel` which takes a `make_distribution_fn` function returning the correct distribution for that step of the MCMC chain?
2. Fair point.  I'll give this some careful thought -- probably along the lines of Fonnesbeck 2013 (which STAN uses a marginalised implementation for, but as far as I understand the original implementation mixed over a discrete space). 
3. (and 6) I guess this one's up to how the TFP semantics evolve.  For a `JointDistribution*`, the state parts map onto stochastic nodes in the respective DAG describing the model, which I think this makes sense for most applications.  The exception would be on-the-fly non-centering of models (e.g. [Xiao Li Meng's ASIS algorithm](https://www.tandfonline.com/doi/abs/10.1198/jcgs.2011.203main) or some of the partially-non-centred epidemics work such as [Neal and Roberts 2005](https://link.springer.com/article/10.1007/s11222-005-4074-7)).  What's the direction of travel with the semantics?
4. Yeah, couldn't see a way around that, for the same reasons as (5) next...
5. Unwrapping `TransformedTransitionKernel`s.  If I've understood your question correctly... Consider the sequence of Gibbs steps for single-parameter updates of `alpha`, `beta`, and `sigma` in the linear model
```
y = Normal(loc = alpha + beta*x, scale=sigma)
```
`alpha` and `beta` exist on an unconstrained space, whereas `sigma` is strictly positive and must be transformed to an unconstrained space for effective sampling.  The conditional posteriors for all parameters do not factorise into independent terms, so rather than re-calculating the `current_target_log_prob` at each Gibbs step, it's quicker just to recognise the presence of a transformed kernel and add or subtract the Jacobian term as required.  
6. (see 3)
7. `GibbsKernelResults.inner_results` is of type `list` (of inner `TransitionKernel` results namedtuples).  At the time when we wrote the code, the `unnest` recursive descent algorithm wasn't able to iterate over a list structure when it met it, and so threw an exception.  Not sure if this has changed?   I guess one possibility might be to dynamically construct a `namedtuple` for `GibbsKernelResults` which replaces the `list` with a bunch of fields. 

We'll do the easy bits first, and address (2) then (1).  

Maybe I could ask the team for their thoughts around how `state_parts` semantics might play into the future of this approach?  It seems a good time to talk about it with the commit of @ColCarroll's turnkey windowed sampling solution (which AFAIU uses restructuring and shape bijectors to concatenate state parts).

Reply:
> presumably only calibrated as long as each make_kernel_fn is also calibrated?

And yes, we should try to instantiate each kernel by calling each `make_kernel_fn` and take logical AND of all `is_calibrated`